example_project/mappers/residential/samples/util.rb

```diff
@@ -3,13 +3,237 @@
+      acc && current_match
+    end
+
+    #puts "IS_MATCH = #{is_match}" #for debugging
+
+    if is_match
+      # "Building" is the building id header in buildstock csv
+      matches << row['Building']
+    end
+  end
+
+  ## check the matches we got and select from them
+  case matches.size
+  when 0
+    infos << "\n Feature #{feature_id}: No matching buildstock building ID found. #{mapped_properties}"
+    selected_id = nil
+  when 1
+    selected_id = matches.first
+    infos << "\n Feature #{feature_id}: Matching buildstock building ID found: #{selected_id}. #{mapped_properties}"
+  else
+    selected_id = matches.sample
+    infos << "\n Feature #{feature_id}: Multiple matches found. Selected one buildstock building ID randomly: #{selected_id} from #{matches.size} matching buildings: #{matches}. #{mapped_properties}"
+  end
+
+  ### Log matching results to csv 
+  # Path to your log CSV file
+  log_csv_path = File.join(File.dirname(__FILE__), '../../../run/uo_buildstock_match_log.csv')
+
+  full_path = File.join(Dir.pwd, log_csv_path)
+  puts "CSV is saved at: #{full_path}"
+
+  # Initialize CSV file with headers if it doesn't exist
+  unless File.exist?(log_csv_path)
+    CSV.open(log_csv_path, 'wb') do |csv|
+      csv << ['uo_id', 'matches', 'selected_match']
+    end
+  end
+  # Append data to matching data CSV file
+  CSV.open(log_csv_path, 'ab') do |csv|
+    # Convert matches array to a string to store in CSV
+    matches_str = matches.join('; ')
+    selected_match_str = selected_id.to_s # selected_id is the result of the case statement
+
+    # Append new row with feature ID, matches, and selected match
+    csv << [feature_id, matches_str, selected_match_str]
+  end
+
+  #Return selected_id, infos
+  return selected_id, infos
+end
+
+def map_to_resstock_building_type(res_type, number_of_residential_units)
+  '''Define function to map building type to categories'''
+  
+  if res_type == 'Multifamily'
+    if number_of_residential_units >= 5
+      return 'Multi-Family with 5+ Units'
+    elsif number_of_residential_units.between?(2, 4)
+      return 'Multi-Family with 2 - 4 Units'
+    end              
+  elsif res_type == 'Single-Family Attached'
+    return 'Single-Family Attached'
+  elsif res_type == 'Single-Family Detached'
+    return 'Single-Family Detached'
+  elsif res_type == 'Mobile Home'
+    return 'Mobile Home'
+  else 
+    return 'Other Category'
+  end
+
+end
+
+def map_to_resstock_floor_area(floor_area, number_of_residential_units)
+  '''floor area mapping using "Geometry Floor Area" floor area'''
+
+  floor_area_mapping = {
+    '750-999' => [750, 999],
+    '1000-1499' => [1000, 1499],
+    '500-749' => [500, 749],
+    '3000-3999' => [3000, 3999],
+    '2000-2499' => [2000, 2499],
+    '1500-1999' => [1500, 1999],
+    '2500-2999' => [2500, 2999],
+    '4000+' => [4000, Float::INFINITY],
+    '0-499' => [0, 499]
+  }
+  resstock_floor_area = nil
+
+  floor_area /= number_of_residential_units
```

Comment:
Feature floor area describes the whole building, not just the unit floor area like ResStock does.